src/ray/core_worker/experimental_channel_manager.h

```diff
@@ -0,0 +1,146 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "ray/common/buffer.h"
+#include "ray/common/ray_object.h"
+#include "ray/common/status.h"
+#include "ray/object_manager/plasma/client.h"
+#include "ray/object_manager/plasma/common.h"
+#include "ray/object_manager/plasma/plasma.h"
+#include "ray/util/visibility.h"
+#include "src/ray/protobuf/common.pb.h"
+
+namespace ray {
+
+class ExperimentalChannelManager {
+ public:
+  ExperimentalChannelManager() {}
+
+  /// Register the caller as a writer for the channel.
+  ///
+  /// \param[in] object_id The ID of the object.
+  /// \param[in] mutable_object Struct containing pointers for the object
+  /// header, which is used to synchronize with other writers and readers, and
+  /// the object data and metadata, which is read by the application.
+  /// \return The return status.
+  Status RegisterWriterChannel(const ObjectID &object_id,
+                               std::unique_ptr<plasma::MutableObject> mutable_object);
+
+  /// Acquires a write lock on the object that prevents readers from reading
+  /// until we are done writing. This is safe for concurrent writers.
+  ///
+  /// \param[in] object_id The ID of the object.
+  /// \param[in] data_size The size of the object to write. This overwrites the
+  /// current data size.
+  /// \param[in] metadata A pointer to the object metadata buffer to copy. This
+  /// will overwrite the current metadata.
+  /// \param[in] metadata_size The number of bytes to copy from the metadata
+  /// pointer.
+  /// \param[in] num_readers The number of readers that must read and release
```

Comment:
I am a little confused about this arguments. Does this mean "the number of readers that must read and release to acquire a write lock again?"

Reply:
Also, what's happening if WriteRelease if the object is not read as many as num_readers?